include/nuttx/ipcc.h

```diff
@@ -0,0 +1,188 @@
+ * Such instance is passed to ipcc_register() function after driver
+ * has been initialized, binding the upper and lower halves into one
+ * driver.
+ */
+
+struct ipcc_lower_s
+{
+  FAR struct pollfd        *fds[CONFIG_IPCC_NPOLLWAITERS];
+  FAR struct ipcc_ops_s     ops;      /* Arch specific functions */
+  sem_t                     rxsem;    /* Data ready to read semaphore */
+  sem_t                     txsem;    /* Data ready to send semaphore */
+  int                       chan;     /* IPCC channel */
+
+#ifdef CONFIG_IPCC_BUFFERED
+  struct circbuf_s          rxbuf;    /* Receive buffer */
+  struct circbuf_s          txbuf;    /* Transmit buffer */
+
+  /* 1- not all data could fit into buffer, some unread data is still
+   * in IPCC mbox. State of this variable is maintained by lower half,
+   * upper half only reads it to know when to inform lower driver that
+   * buffer got free.
+   */
+
+  int                       overflow;
+#endif /* CONFIG_IPCC_BUFFERED */
+};
+
+/* Driver state structure */
+
+struct ipcc_driver_s
+{
+  FAR struct ipcc_lower_s   ipcc;     /* Lower half driver state */
+  int                       crefs;    /* Count number of open references */
+  int                       unlinked; /* 1 - driver has been unlinked */
+  sem_t                     exclsem;  /* Mutual exclusion for driver */
+};
+
+/****************************************************************************
+ * Public Functions Prototypes
+ ****************************************************************************/
+
+void ipcc_pollnotify(FAR struct ipcc_lower_s *ipcc, pollevent_t eventset);
+void ipcc_cleanup(FAR struct ipcc_driver_s *priv);
+void ipcc_txfree_notify(FAR struct ipcc_lower_s *priv);
+void ipcc_rxfree_notify(FAR struct ipcc_lower_s *ipcc);
+
+#ifdef CONFIG_IPCC_BUFFERED
+int ipcc_register(int chan, size_t rxbuflen, size_t txbuflen);
+#else
+int ipcc_register(int chan);
+#endif
+
+/****************************************************************************
+ * Name: up_ipcc_init
+ *
+ * Description:
+ *   Initializes lower half of the driver. This function should be
+ *   implemented by lower half driver in microcontroller code. Function
+ *   should not initialize ipcc_lower_s structure with the exception of
+ *   explicitly specified fields like setting pointer to read and write
+ *   function pointers. Rest of the fields are initialized in upper half
+ *   driver.
+ *
+ *   This function is called once in ipcc_register() function.
+ *
+ * Input Parameters:
+ *   ipcc - interface between upper and lower half drivers. Only some
+ *          of the fields should be initialized in this function.
+ *   chan - channel number to initialize.
+ *
+ * Returned Value:
+ *   0 on success or negated errno otherwise.
+ *
+ * Assumptions/Limitations:
+ *
+ ****************************************************************************/
+
+int up_ipcc_init(FAR struct ipcc_lower_s *ipcc, int chan);
```

Comment:
up_xxx is reserved to the core os interact with the arch/chip specific code, it isn't good to use up api for the normal driver communication.

Reply:
Ok, sure, I'll change this.

Reply:
ipcc_lower_init() maybe? and ipcc_lower_cleanup() for cleanup function?

Reply:
Ok, more work for the user but it makes sense. Will change that.